Auto branch docs

pp-mo opened the pull request:
Closes #3396 

NOTE: I have made version changes (2.2 --> 2.3)  to the .RSTs in 3 places :
  * `docs/iris/src/_templates/index.html`
  * `docs/iris/src/_templates/layout.html`
  * `docs/iris/src/whatsnew/index.rst`

In future, to make this scheme work properly (so branch docs are properly readable, especially the whats-new), we really need to move these on _just after a new release, when we bump the version in master_.
That is, exactly as we do to the iris version string.

lbdreyer commented on .travis.yml:
```diff
@@ -135,6 +135,25 @@ script:
+      WHATSNEW=$(ls -d contributions_* 2>/dev/null);
+      if [[ "$WHATSNEW" != "" ]]; then
+        VERSION=$(echo $WHATSNEW | grep -o "[[:digit:].]*");
+        python aggregate_directory.py $VERSION;
```
I think aggregate_directory should also take an optional flag to set the release date to be something like "Unreleased" or "TBC"
Currently, it [uses the current date](https://github.com/SciTools/iris/blob/7419f8839510d8b784bcca9567fe5ac9a9aaa68a/docs/iris/src/whatsnew/aggregate_directory.py#L180) so if you look at the [Iris 2.3 whats new](https://scitools-docs.github.io/iris/auto_branch_docs/whatsnew/2.3.html) that is generated, it claims that it has already been released...
I think this will just lead to confusion

pp-mo replied:
ok

lbdreyer commented on .travis.yml:
```diff
@@ -135,6 +135,25 @@ script:
+      python aggregate_directory.py --checkonly;
+    fi
+
+  # When pushing built docs, attempt to make a preliminary whatsnew with 'aggregate_directory.py', before the build.
```
This line is a bit long. Could you split it over two lines to make it more readable

lbdreyer commented on README.md:
```diff
@@ -85,6 +85,11 @@ use of standard NumPy/dask arrays as its underlying data storage.
+
+A documentation build for the latest development code is visible at
+<https://scitools-docs.github.io/iris/master/index.html>.
+
```
In the past we also had a badge in the [README that pointed to the docs built from master](https://github.com/SciTools/iris/blob/962f0159513e6eb08ed1a76fbc8485a31dd584bb/README.md). Could you readd it?

pp-mo replied:
ok!

lbdreyer commented on README.md:
```diff
@@ -85,6 +85,11 @@ use of standard NumPy/dask arrays as its underlying data storage.
 The documentation for Iris is available at <https://scitools.org.uk/iris/docs/latest>,
 including a user guide, example code, and gallery.
 
+## Work in progress
```
Could we remove this title?
To me, it reads as if this section is a WIP in the sense that at some point we hope for it not to be "in progress anymore" and to actually finalise on the contents. But that's not what will happen. It will always be "in progress".

I think it would be fine if 

> A documentation build for the latest development code is visible at https://scitools-docs.github.io/iris/master/index.html.

is part of the section above

pp-mo replied:
ok!

lbdreyer replied:
@pp-mo It looks like this title is still yet to be removed.

pp-mo replied:
Sorry, that change got lost when I reverted something.
Will reinstate...

pp-mo commented:
Thanks @lbdreyer , I hope I've addressed those points now.
Mostly : I have added a 'special' mode to the aggregation as suggested, and it produces a temporary whatsnew page with a temporary non-version name.
See : https://scitools-docs.github.io/iris/auto_branch_docs/whatsnew/index.html

lbdreyer commented:
Looking good! Thanks @pp-mo !
